docs/proposals/improved-schema-for-services.md

```diff
@@ -0,0 +1,442 @@
+Getting that metadata from the cluster is challenging because a normal cluster user ( plain vanilla ) doesn't have access to that metadata so we are gonna follow a 3 tiered approach
+
+- check if the user has access to the CRD, if so then use that.
+- get the `swagger.json` from the openapi endpoint provided by kubernetes. a sample can be found here https://gist.github.com/girishramnani/f29949d5cb8c6547102776437e05ac19
+- finally if we cannot get the openapi data as well then use the `ClusterServiceVersion` to get the parameters.
+
+This change would affect multiple service commands and the changes are described briefly below -
+- `odo catalog describe service` should include the metadata fields with description so the users can provide then when doing `odo service create`. 
+- `odo catalog list service` shouldn't change much other then listing CRDs per operator
+- `odo service create` should take flags and dynamically fill the CRD structs - metadata would be used for validation
+- `odo service create --from-file` would be used by adapter team or whoever wants to provide the whole CRD themselves as a file. this feature is already present and working.
+- `json` output for all commands.
+
+# Implementation plan
+
+Note - Below I would mention `cluster` manytimes and that means both openshift and kuberenetes. If there is something specific to one or the other then I would mention it.
+
+Below is the step wise plan for implementing this feature - 
+
+## Getting the metadata
+
+### Admin Access/Access to CRD API path
+First we would consider the scenerio where the user has admin access or privileges to access the CRD API to the cluster.
+Cluster provides an api to get metadata for any CRD needed. URL - `<cluster-url>/api/kubernetes/apis/apiextensions.k8s.io/v1/customresourcedefinitions`. 
+
+Sample output - 
+https://gist.github.com/girishramnani/cbb4400e463efe89c13f1386e0788793
+
+We care about the `openAPIV3Schema` as that would be used to build the CRD struct. 
+
+### Kuberenetes cluster Swagger has the schema
+
+If the user doesn't have CRD access then we fetch try to fetch the same `openAPIV3Schema` from the cluster's `swagger.json` from the endpoint `<cluster-url>/api/kubernetes/openapi/v2`. This is a very large document as it holds all the definitions present on the cluster.
+
+So this needs to be cached and refreshed whenever a new operator is installed.
+
+The caching would be per cluster as the `swagger.json` would change for different clusters. The approach would be as follows -
+
+- when the users executes `odo catalog list services`, if there is no cache then we would fetch the `swagger.json`, odo wouldn't exit until download. 
```

Comment:
> odo wouldn't exit until download

As you mentioned on contributor's call, we had a similar situation with Telemetry. I think this one is worth exploring options where we can do things in background without blocking an odo command. I don't mean to do it in the first pass, but eventually.

I think it's worth it because `odo catalog list services` is going to be used extensively as we pitch odo more and more as a tool that plays well with services. And every time a new Operator is installed, the process of downloading `swagger.json` is going to be repeated. Although these are presumptions and not backed by Telemetry.

Reply:
Well I don’t have a problem with either approach but what if the file is not downloaded and the user tries to create a service, it might be confusing

Reply:
@kadel can we conclude on the approach here?

Reply:
> Well I don’t have a problem with either approach but what if the file is not downloaded and the user tries to create a service, it might be confusing

We can print a message at the end of `odo catalog list services` saying that the information is being downloaded from the cluster and that the user should wait for a few seconds.

Besides `odo catalog list services`, don't you think we will need to validate our cache for `odo catalog describe service <operator>/<cr>` command as well?

Reply:
The validation which comes to my mind is mostly about it being a valid json.
Because validating anything else might be equal to downloading the file from cluster everytime.

Reply:
> We can print a message at the end of odo catalog list services saying that the information is being downloaded from the cluster and that the user should wait for a few seconds.

then why not just block on the command?

Reply:
Let's start with simply blocking the command. Operators are not installed that often, so most of the time the cache will be used.
Let's just make sure that users know what is going on, and what they are waiting for. 
Something like "Downloading service information from the cluster ........"

Reply:
ack, adding that ask a comment

Reply:
done